### PR TITLE
Add JWT Grant authentication

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20
| License       | MIT

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,18 +64,22 @@ docusign:
...
+[Configure JWT Grant](https://developers.docusign.com/esign-rest-api/guides/authentication/oauth2-jsonwebtoken).
 Your account id is visible on the top right level of your demo.docusign account right below your profile picture in the little drop-down.
 
 ```yml
 # config/packages/docusign.yml
 
 docusign:
-    access_token: "YourAccessToken"
+    demo: "%kernel.debug"
+    jwt:
```
Comment (GregoireHebert) [inline README.md]: `auth_jwt` ? 😄

File: tests/DependencyInjection/ConfigurationTest.php [hunk 2]
```diff
@@ -33,7 +33,12 @@ public function testDefaultConfig(): void
         $treeBuilder = $this->configuration->getConfigTreeBuilder();
         $config = $this->processor->processConfiguration($this->configuration, [
             'docusign' => [
-                'access_token' => 'token',
+                'demo' => false,
```
Comment (GregoireHebert) [inline tests/DependencyInjection/ConfigurationTest.php]: Would you mind adding a `'demo' => true,` test over the 2 concerned configuration keys?
I saw a test configuration with it but I did not see any test over the expected result.

File: README.md [hunk 3]
```diff
@@ -64,18 +64,22 @@ docusign:
 ### Configure the bundle
 
 Check the [official documentation](https://github.com/docusign/qs-php).
-[Get your testing access token](https://developers.docusign.com/oauth-token-generator).
+[Configure JWT Grant](https://developers.docusign.com/esign-rest-api/guides/authentication/oauth2-jsonwebtoken).
```
Comment (GregoireHebert) [inline README.md]:
```suggestion
[Configure JWT Grant](https://developers.docusign.com/esign-rest-api/guides/authentication/oauth2-jsonwebtoken) and set your key into `%kernel.project_dir%/var/jwt/docusign.pem`.
```

File: src/Grant/JwtGrant.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    public function __invoke(): string
+    {
+        $time = time();
+        $token = (new Builder())->issuedBy($this->integrationKey) // iss
+            ->relatedTo($this->userGuid) // sub
+            ->issuedAt($time) // iat
+            ->expiresAt($time + 3600) // exp
+            ->permittedFor('account-d.docusign.com') // aud
```
Comment (GregoireHebert) [inline src/Grant/JwtGrant.php]: Shouldn't this come from thee `docusign.account_api_uri` parameter?

File: src/Grant/JwtGrant.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        $this->accountApiUri = $accountApiUri;
+    }
+
+    public function __invoke(): string
+    {
+        $time = time();
+        $token = (new Builder())->issuedBy($this->integrationKey) // iss
+            ->relatedTo($this->userGuid) // sub
+            ->issuedAt($time) // iat
+            ->expiresAt($time + 3600) // exp
```
Comment (GregoireHebert) [inline src/Grant/JwtGrant.php]: IMO this should be a configuration option. In some cases I might need to reduce this value.

File: src/DependencyInjection/DocusignExtension.php [hunk 6]
```diff
@@ -18,6 +18,9 @@
 final class DocusignExtension extends Extension
 {
     public const STORAGE_NAME = 'docusign.storage';
+    private const DEMO_API_URI = 'https://demo.docusign.net/restapi';
+    private const DEMO_ACCOUNT_API_URI = 'https://account-d.docusign.com/oauth/token';
+    private const ACCOUNT_API_URI = 'https://account.docusign.net/oauth/token';
```
[the remark this answers is not in the document]
Comment (vincentchalamon) [reply src/DependencyInjection/DocusignExtension.php]: On production, it's always `account.docusign.net`, so it's not configurable

File: src/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -52,26 +52,26 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->end()
                 ->end()
                 ->scalarNode('api_uri')
-                    ->info('Docusign API URI (uses the demo by default)')
+                    ->info('DocuSign production API URI (default: https://www.docusign.net/restapi)')
                     ->cannotBeEmpty()
-                    ->defaultValue('https://demo.docusign.net/restapi')
+                    ->defaultValue('https://www.docusign.net/restapi')
```
[the remark this answers is not in the document]
Comment (vincentchalamon) [reply src/DependencyInjection/Configuration.php]: The validate process is built to validate a data, not to change (force) it

File: tests/DependencyInjection/ConfigurationTest.php (hunk 2 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/ConfigurationTest.php]: Test has been done in DocusignExtensionTest with demo mode

**Comment by GregoireHebert** [comment]
Thanks @vincentchalamon this is nice :D

**Pull request merged**